x/session/keeper/session_hydrator.go

```diff
@@ -0,0 +1,203 @@
+	sessionHeader *types.SessionHeader
+
+	// The fully hydrated session object
+	session *types.Session
+
+	// The height at which the session being request
+	blockHeight int64
+
+	// A redundant helper that maintains a hex decoded copy of `session.Id` used for session hydration
+	sessionIdBz []byte
+}
+
+func NewSessionHydrator(
+	appAddress string,
+	serviceId string,
+	blockHeight int64,
+) *sessionHydrator {
+	sessionHeader := &types.SessionHeader{
+		ApplicationAddress: appAddress,
+		ServiceId:          &sharedtypes.ServiceId{Id: serviceId},
+	}
+	return &sessionHydrator{
+		sessionHeader: sessionHeader,
+		session:       &types.Session{},
+		blockHeight:   blockHeight,
+		sessionIdBz:   make([]byte, 0),
+	}
+}
+
+// GetSession implements of the exposed `UtilityModule.GetSession` function
+// TECHDEBT(#519): Add custom error types depending on the type of issue that occurred and assert on them in the unit tests.
+func (k Keeper) HydrateSession(ctx sdk.Context, sh *sessionHydrator) (*types.Session, error) {
+	logger := k.Logger(ctx).With("method", "hydrateSession")
+
+	if err := k.hydrateSessionMetadata(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session metadata: %v", err)
+	}
+	logger.Debug("Finished hydrating session metadata")
+
+	if err := k.hydrateSessionID(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate the session ID: %v", err)
+	}
+	logger.Info("Finished hydrating session ID: %s", sh.sessionHeader.SessionId)
+
+	if err := k.hydrateSessionApplication(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate application for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session application: %+v", sh.session.Application)
+
+	if err := k.hydrateSessionSuppliers(ctx, sh); err != nil {
+		return nil, sdkerrors.Wrapf(types.ErrHydratingSession, "failed to hydrate suppliers for session: %v", err)
+	}
+	logger.Debug("Finished hydrating session suppliers: %+v")
+
+	sh.session.Header = sh.sessionHeader
+	sh.session.SessionId = sh.sessionHeader.SessionId
+
+	return sh.session, nil
+}
+
+// hydrateSessionMetadata hydrates metadata related to the session such as the height at which the session started, its number, the number of blocks per session, etc..
+func (k Keeper) hydrateSessionMetadata(ctx sdk.Context, sh *sessionHydrator) error {
+	sh.session.NumBlocksPerSession = NumBlocksPerSession
+	sh.session.SessionNumber = int64(sh.blockHeight / NumBlocksPerSession)
+	sh.sessionHeader.SessionStartBlockHeight = sh.blockHeight - (sh.blockHeight % NumBlocksPerSession)
+	return nil
+}
+
+// hydrateSessionID use both session and on-chain data to determine a unique session ID
+func (k Keeper) hydrateSessionID(ctx sdk.Context, sh *sessionHydrator) error {
+	// TODO_TECHDEBT: Need to retrieve the block hash at SessionStartBlockHeight, NOT THE CURRENT ONE
+	prevHashBz := ctx.HeaderHash()
```

Comment:
IMO, for the stake of having stable `SessionId`s, we should just omit `prevHashBz` for now

Reply:
`prevHashBz` will get the `HeaderHash` of the current block (time when the query is performed). So if I query GetSession at block 10 or 11 `prevHashBz` would be different, which makes the seessionId different and dependent on the current block height.

Reply:
Discussed offline. 

We are temporarily going to stop use `headerHash` because it will cause inconsistencies during development but marking it as a TODO_BLOCKER.

![Screenshot 2023-10-26 at 3 15 19 PM](https://github.com/pokt-network/poktroll/assets/1892194/c2e58fdf-ce94-4c79-917a-f44a920119fa)